### PR TITLE
Add transformers to get features from raw data. Update pipeline.

### LINKED ISSUE #380: Update time-to-merge Service to Accept Raw Data as Input
Currently the deployed service (#348) takes feature engineered PR data as input, and yields the estimated time to merge of the PR as output. However, if this service is to be used by thoth / openshift-ci / other github bots, then it would probably only have the raw PR data (such as that collected via [mi-scheduler](https://github.com/thoth-station/mi-scheduler)) available. And therefore, it won't be able to yield any meaningful output.

One idea to remedy this issue is to update the deployed Pipeline object in the service to include feature engineering objects as well.

Related to #236, #362

Acceptance Criteria
- [x] feature engineering steps added to pipeline in deployed service

### PR DESCRIPTION
**Pull request by chauhankaranraj**
## Related Issues and Dependencies

Closes #380 

## This introduces a breaking change

- [ ] Yes
- [x] No

## This Pull Request implements

- Takes the feature engineering methods explored and found to be the most effective from the EDA notebook, and wraps them into various custom Preprocessor/Transformer objects
- Updates the prediction pipeline to take raw data as input, and use this transformer objects to get feature-engineered data
- Saves the updated pipeline to s3 bucket

### CODE REVIEW

**Comment by chauhankaranraj** [comment]
Putting this on /hold /do-not-merge - seldon's pre-packaged sklearn inference server does not seem to work well with custom transformers. So we may have to look for a workaround, and ensure that works with the custom transform logic

**Review by Shreyanand — COMMENTED** [review]
Looks good! just added a discussion point.

**Comment by chauhankaranraj** [comment]
> Perhaps raw_data_processor.get_feature_names()?

Hrm that seems to throw an error - is this not the correct way to use it?

Screenshot from 2021-09-22 15-13-54

**Comment by Shreyanand** [comment]
> Yep that was pretty much the guiding principle when writing these classes :) For example, one of the model inputs is PR body character count. Now if in addition to PR body we also wanted title or comments in the future, we can just use the same `StringLenTransformer` for these columns as well. Another example is `DateTimeDetailsTransformer`. However columns like num. files of each extension changed (`FileTypeCounter`) seemed difficult to generalize and therefore had to be designed to work on just 1 column.

That makes sense! One way of generalizing could be to have `UserAttributes` with logic from `IsUserPrivTransformer`, and `NumPrevPRsTransformer`. `PRAttributes` with logic from `TitleWordCountTransformer` , `StringLenTransformer`,  `pr_size_encoder`, and `DateTimeDetailsTransformer`. Finally, `PRFilesAttributes` with logic from `ChangeInDirTransformer`, `NumChangedFilesTransformer`, and `FileTypeCountTransformer`.

**Comment by Shreyanand** [comment]
> > Perhaps raw_data_processor.get_feature_names()?
> 
> Hrm that seems to throw an error - is this not the correct way to use it?
> 
> Screenshot from 2021-09-22 15-13-54

huh, I guess it doesn't work with all the [transformers](https://github.com/scikit-learn/scikit-learn/blob/2beed5584/sklearn/compose/_column_transformer.py#L350).

File: src/features/build_features.py [hunk 1]
```diff
@@ -1 +1,352 @@
...
+    ):
+        """Initialize."""
+        self.priv, self.org, self.repo, self.branch = priv, org, repo, branch
+
+        # read and parse owners file from repo
+        ret = requests.get(
+            f"https://raw.githubusercontent.com/{self.org}/{self.repo}/{self.branch}/OWNERS"
+        )
+        content = ret.content.decode("utf-8")
+        self.owners_file = yaml.load(content, Loader=yaml.SafeLoader)
```
[the remark this answers is not in the document]
Comment (chauhankaranraj) [reply src/features/build_features.py:128]: Cool, updated :smile:

**Comment by chauhankaranraj** [comment]
> @chauhankaranraj can you rebase off of current master? :) then we'll merge

Rebased :)

**Comment by chauhankaranraj** [comment]
@harshad16 I can't ascertain why the build-check is failing on this PR, any ideas?

**Comment by harshad16** [comment]
> I can't ascertain why the build-check is failing on this PR, any ideas?

oh the scr is what is breaking this:
2021-11-12 18:17:24,427 [806] INFO     micropipenv: Installing '\''src'\''
ERROR: Could not find a version that satisfies the requirement src==0.1.0 (from versions: 0.0.1, 0.0.3, 0.0.4, 0.0.5, 0.0.6, 0.0.7)
ERROR: No matching distribution found for src==0.1.0

**Comment by oindrillac** [comment]
Seems like https://github.com/thoth-station/micropipenv/pull/204 is merged.

/retest

**Comment by oindrillac** [comment]
> Seems like [thoth-station/micropipenv#204](https://github.com/thoth-station/micropipenv/pull/204) is merged.
> 
> /retest

that did not seem to trigger the retest. Are thoth services down due to the smaug outage? 
https://github.com/operate-first/SRE/issues/430

**Comment by chauhankaranraj** [comment]
All the tests seem to be passing now, could we have a final review please :smile:
